node/src/types/block.rs

```diff
@@ -1834,6 +1834,73 @@ impl Item for BlockWithMetadata {
     }
 }
 
+#[derive(DataSize, Clone, Debug, PartialOrd, Ord, PartialEq, Eq, Hash, Serialize, Deserialize)]
+/// Wrapper around block and its deploys.
+pub struct BlockAndDeploys {
+    /// Block part.
+    pub block: Block,
+    /// Deploys.
+    pub deploys: Vec<Deploy>,
+}
+
+impl Display for BlockAndDeploys {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "block {} and deploys", self.block.hash())
+    }
+}
+
+impl Item for BlockAndDeploys {
+    type Id = BlockHash;
+
+    type ValidationError = BlockValidationError;
+
+    const TAG: Tag = Tag::BlockAndDeploysByHash;
+
+    // false b/c we're not validating finality signatures.
+    const ID_IS_COMPLETE_ITEM: bool = false;
+
+    fn validate(
+        &self,
+        verifiable_chunked_hash_activation: EraId,
+    ) -> Result<(), Self::ValidationError> {
+        let _ = self.block.verify(verifiable_chunked_hash_activation)?;
+        // Validate that we've got all of the deploys we should have gotten, and that their hashes
+        // are valid.
+        for deploy_hash in self
+            .block
+            .deploy_hashes()
+            .iter()
+            .chain(self.block.transfer_hashes().iter())
+        {
+            match self
+                .deploys
+                .iter()
+                .find(|&deploy| deploy.id() == deploy_hash)
+            {
```

Comment:
What is the relationship between `block.deploy_hashes()` + `block.transfer_hashes()` compared to `deploys`? I see here we're checking that all the deploy hashes in block are also in deploys, but are we supposed to check the inverse as well? Also, I see we can receive this from a peer as well, which I assume is not trusted. Do we have a checks for those deploys belonging to the block or is that covered in `block.verify()` and `deploy.has_valid_hash()`?

Reply:
> What is the relationship between `block.deploy_hashes()` + `block.transfer_hashes()` compared to `deploys`? ... but are we supposed to check the inverse as well?

Transfers are just a special case of Deploys.  In most places we don't distinguish between transfers and non-transfer deploys, but we do in the block.  In this specific case, we expect the combined list of deploy and transfer hashes in the block to be the same as the set of deploys provided in `BlockAndDeploys`.  I think you're right - we should be checking that there aren't extra deploys being provided.  Good catch!

> Do we have a checks for those deploys belonging to the block or is that covered in block.verify() and deploy.has_valid_hash()?

Yes, those checks cover the case of proving the deploy belongs in the given block.

Reply:
Fixed in [ecf4e89](https://github.com/casper-network/casper-node/pull/3010/commits/ecf4e895ee12a1b1c67900fe7d519bc4ed9dc658).